### PR TITLE
config: mark `block_sync` key as deprecated

### PR DESCRIPTION
**Pull request by hvanz**
Part of #409

---

#### PR checklist

- [x] Tests written/updated
- [x] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [x] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

File: test/e2e/pkg/manifest.go [hunk 1]
```diff
@@ -121,6 +121,7 @@ type ManifestNode struct {
 
 	// BlockSync specifies the block sync mode: "" (disable), "v0" or "v2".
 	// Defaults to disabled.
+	// Note that BlockSync will be enabled by default in the next major release.
```
Comment (sergio-mena) [inline test/e2e/pkg/manifest.go:124]:
```suggestion
	// Note that BlockSync will always be enabled in the next major release.
```

File: .changelog/unreleased/breaking-changes/409-deprecate-blocksyncmode [hunk 2]
```diff
@@ -0,0 +1,5 @@
+- [config] The boolean key `fastsync` is deprecated and replaced by
+    `block_sync`. ([\#9259](https://github.com/tendermint/tendermint/pull/9259))
+    At the same, `block_sync` is also deprecated, as BlocSync will be enabled by
```
Comment (sergio-mena) [inline .changelog/unreleased/breaking-changes/409-deprecate-blocksyncmode:3]:
```suggestion
    At the same time, `block_sync` is also deprecated, as BlocSync will be enabled by
```

File: .changelog/unreleased/breaking-changes/409-deprecate-blocksyncmode [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
+    default in the next release.
```
Comment (sergio-mena) [inline .changelog/unreleased/breaking-changes/409-deprecate-blocksyncmode:4]:
```suggestion
    At the same time, `block_sync` is also deprecated, as BlocSync will always be enabled
    in the next release.
```

File: config/config.go [hunk 4]
```diff
@@ -165,6 +165,9 @@ func (cfg *Config) CheckDeprecated() []string {
 	if cfg.Mempool.TTLDuration != 0 {
 		warnings = append(warnings, "prioritized mempool key detected. This key, together with this version of the mempool, will be removed in the next major release.")
 	}
+	if !cfg.BaseConfig.BlockSyncMode {
+		warnings = append(warnings, "disabled block_sync key detected. BlockSync will be enabled by default in the next major release and this key will be removed.")
+	}
```
Comment (sergio-mena) [inline config/config.go:170]:
```suggestion
		warnings = append(warnings, "disabled block_sync key detected. BlockSync will be enabled unconditionally in the next major release and this key will be removed.")
	}
```

File: config/config.go [hunk 5]
```diff
@@ -196,6 +199,7 @@ type BaseConfig struct { //nolint: maligned
 	// If this node is many blocks behind the tip of the chain, Blocksync
 	// allows them to catchup quickly by downloading blocks in parallel
 	// and verifying their commits
+	// Deprecated: BlockSync will be enabled by default in the next major release.
```
Comment (sergio-mena) [inline config/config.go:202]:
```suggestion
	// Deprecated: BlockSync will be enabled unconditionally in the next major release.
```

File: config/toml.go [hunk 6]
```diff
@@ -90,6 +90,7 @@ moniker = "{{ .BaseConfig.Moniker }}"
 # If this node is many blocks behind the tip of the chain, BlockSync
 # allows them to catchup quickly by downloading blocks in parallel
 # and verifying their commits
+# Deprecated: BlockSync will be enabled by default in the next major release.
```
Comment (sergio-mena) [inline config/toml.go:93]:
```suggestion
# Deprecated: BlockSync will be enabled unconditionally in the next major release.
```

File: docs/core/configuration.md [hunk 7]
```diff
@@ -39,6 +39,7 @@ moniker = "anonymous"
 # If this node is many blocks behind the tip of the chain, BlockSync
 # allows them to catchup quickly by downloading blocks in parallel
 # and verifying their commits
+# Deprecated: BlockSync will be enabled by default in the next major release.
```
Comment (sergio-mena) [inline docs/core/configuration.md:42]:
```suggestion
# Deprecated: BlockSync will be enabled unconditionally in the next major release.
```

**Review by sergio-mena — APPROVED** [review]

**Comment by hvanz** [comment]
@sergio-mena Thanks for the suggestions!

**Comment by cason** [comment]
As a general comment here, do you think most users will understand what we mean with this messages? Maybe we should make clear that the flag (and section) about blocksync will be removed...

**Comment by hvanz** [comment]
@cason We decided not to remove the BlockSync section and the version field because we will add more versions in the future. We are just removing the flag, for which there enough messages for the users as log warnings, comments, and in the changelog. I agree though that is not the best to rename a flag and remove it all in the same release.

**Comment by cason** [comment]
Hey, I was more considering the wording of the message. I would be for a more direct message, such as: "Block Sync cannot be disabled", instead of "BlockSync will be enabled unconditionally".

**Comment by hvanz** [comment]
@cason I made it more explicit in the messages that we are removing the key.

File: test/e2e/pkg/manifest.go [hunk 8]
```diff
@@ -121,6 +121,9 @@ type ManifestNode struct {
 
 	// BlockSync specifies the block sync mode: "" (disable), "v0" or "v2".
 	// Defaults to disabled.
+	//
+	// Note that BlockSync will always be enabled in the next major release the
```
Comment (cason) [inline test/e2e/pkg/manifest.go:125]:
```suggestion
	// Note that BlockSync will always be enabled in the next major release and the
```

**Review by cason — APPROVED** [review]

**Pull request merged**